### ISSUE TITLE
Small change needed for flatpak support

### ISSUE THREAD

**Issue by or29544** [puddletag/puddletag#735]
I am working on publishing puddletag on flathub. The pull request is made and the flatpak manifest and application definitions have passed all checks. The app build and runs, tags are saved and everything. The flathub maintainers though found an issue with how `QFileDialog` is used across the project. To allow `QFileDialog` to use portals and enable the flatpak to communicate with the file system, it needs the `DontUseNativeDialog = true` option everywhere it's used. 

Is this something you would like to enable? It would be great to have puddletag as a flatpak and I am willing to make it happen. If you agree with the `DontUseNativeDialog` I can try to create the pull request myself.

**Comment by sandrotosi** [comment]
@corubba can you comment on this? @or29544 it would be great if you can prepare a PR for the change (making sure the stand-alone app still works as-is and via flatpak), after Corubba approves it -- thanks both!

**Comment by tinywrkb** [comment]
@sandrotosi if you don't set `DontUseNativeDialog` as true, then in Flatpak you're using the [Document Portal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Documents), and this will have a couple of consequences.  

- If you open a folder and then saved as a playlist, the playlist will have full paths for the audio files, and they will be wrong as they will point to `/run/user/UID/doc/...`.
- Loading audio files from a playlist with relative paths is expected to be broken.

I suspect that all file dialogs in the app would need to set `DontUseNativeDialog` as true, at least until there's a [Portal to handle this need](https://github.com/flatpak/xdg-desktop-portal/issues/463).

**Comment by raduzaharia-medium** [comment]
> @sandrotosi if you don't set `DontUseNativeDialog` as true, then in Flatpak you're using the [Document Portal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Documents), and this will have a couple of consequences.
> 
>     * If you open a folder and then saved as a playlist, the playlist will have full paths for the audio files, and they will be wrong as they will point to `/run/user/UID/doc/...`.
> 
>     * Loading audio files from a playlist with relative paths is expected to be broken.
> 
> 
> I suspect that all file dialogs in the app would need to set `DontUseNativeDialog` as true, at least until there's a [Portal to handle this need](https://github.com/flatpak/xdg-desktop-portal/issues/463).

Yup, there are around 15-20 instances of `QFileDialog`. I will add the option to all.

**Comment by sandrotosi** [comment]
i would still like to hear @corubba 's opinion on it

**Comment by raduzaharia-medium** [comment]
Ok, indeed the flathub maintainers were hinting at the exact same issues described by @corubba. I am also linking @tinywrkb from flathub. I am kind of sad because it would have been great to have a flatpak version of puddletag. But I agree, we should not apply a workaround for the sake of an incomplete runtime. I am not really sure how to proceed. It seems the best solution so far is to drop it altogether. 

To be fair, I have no idea why this is not working...the documentation clearly says the KDE/Gnome open file dialog will open a portal when flatpak is detected. Anyway. If anybody can offer a solution for me to implement, I would be more than happy to do it. But again, it is my thought too that `DontUseNativeDialog` is not really what we want or need. And as @sandrotosi said, it might not even work.

**Comment by tinywrkb** [comment]
> o put it bluntly, out of pure principle I am against switching all our users to non-native file-dialogs only to work around the shortcomings of a special runtime environment a handful of users want to use.

No one said this should change for all users. I thought it's pretty obvious that such a change should be behind make configure option.

> I am not even sure where the solution with that flag originates, but flatpak itself recommends not to use it

And it also says

> Note that portals cannot currently give access to directories on the host filesystem

which implies that you should set `DontUseNativeDialog` as true when you want to open a folder. Though I think this is wrong, and documentation is outdated now.

> Alternatively it can also be worked around on the user side by reducing the sandboxing

The application already has filesystem permissions that mounts a few XDG dirs (music, images, and maybe downloads) into the sandbox, so the application has access to them, but that's not the issue.  
The problem is that when you use the native file dialog, you get the document portal and `/run/user/UID/doc/...` paths for files and folders.  

Qt is in the wrong here, forcing native file dialog on apps.  
GTK, on the other hand, handle this as opt-in, and required developers to switch from `Gtk.FileChooserDialog` to `Gtk.FileChooserNative`.

**Comment by raduzaharia-medium** [comment]
Ok then, is everybody good with me enabling DontUseNativeDialog for a new flag? Are any such flags used in puddletag for me to grab for inspiration?

**Comment by sandrotosi** [comment]
> My (admittedly biased) opinion is still to keep this flatpak-specific workaround out of the application and on the flatpak-side of things, and go the patch-during-flatpak-build route until the portal has catched up and native dialogs work in flatpak the same way they do everywhere else. That way these changes are only in the flatpak where they are needed, and we don't have to worry about maintaining the workaround or any negative side-effects on the non-flatpak-users who are still in the majority.

yup i agree with this assessement, in particular with the maintanance burden of this solution that we dont really benefit from nor use.

**Comment by raduzaharia-medium** [comment]
Ok, I have everything I need. I will try to apply a patch for puddletag when creating the flatpak package. Thank you all for technical explanations and contributions :) This issue may be closed at this point as the rest of development will be happening in the flathub area.

**Comment by ssokolow** [comment]
> Note that portals cannot currently give access to directories on the host filesystem

Sounds outdated to me.

[org.freedesktop.portal.FileChooser](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.FileChooser).OpenFile has an optional `directory` boolean in the `options` argument, and org.freedesktop.portal.FileChooser.SaveFiles will display a directory picker.

> Qt is in the wrong here, forcing native file dialog on apps.

Qt was designed as a portable toolkit with a lot of the original target market being on platforms like Windows and macOS where there has been a standard native dialog since the beginning and there were occasionally problems with executive meddling producing "improved" versions of common UI components in third-party apps.

Under those circumstances, it makes sense for it to have originated as "You want to go outside what the platform common dialog APIs allow? You have to justify the cost of constructing a new dialog from scratch."

Given the [API docs for Qt 3.x](https://doc.qt.io/archives/3.3/qfiledialog.html), the option to not use a native dialog on platforms that offer one appears to have been added in Qt 4.

**Comment by raduzaharia-medium** [comment]
Just as a small heads up: flatpak packaging works, but flathub hosting is not happening. I created the patch to add the QFileDialog option, it worked great, I did everything with no disruption to puddletag developers but flathub keep asking to merge my commits upstream and we clearly agreed nobody wants that. It even sounds weird to me: you should not be responsible for every possible runtime showing up now and then and you should not change your source code to cater for everything. The patching approach was the best but the flathub maintainers don't want it. I closed the pull request on their end and won't pursue publishing puddletag there anymore. 

Sorry for the trouble, it would have been nice to have puddletag as a flatpak - I will certainly have it but I will publish it for me privately. I know that flatpak is not everyone's cup of tea but there are legit use cases for it and there are people out there waiting for more flatpak apps to happen. You did everything right on your end: provided technical support, I feel like I did everything right: I implemented the patch and provided clean configuration files, but if flathub want something else, I cannot provide.

**Comment by ssokolow** [comment]
Did you show them this to make the situation clear?

**Comment by raduzaharia-medium** [comment]
Yup, I linked several times to this issue and pointed out the discussion. They keep insisting that flatpak support should be native to the puddletag source code. Which is ridiculous in my opinion.

**Comment by ssokolow** [comment]
Agreed. Given that they allow packaging up things like the Discord client, that's pretty nonsensical.

**Comment by lwbt** [comment]
Flatpak support looks exactly like what I am looking for right now and it's coincidence to find that some else has already done some research and development. But what is happening here in this thread is almost impossible for me to understand. It's no wonder to me why some other party did not agree to proceed with this strategy.

Please help me understand and let me ask constructive questions:
- Where is said patch? Is it [here](https://github.com/puddletag/puddletag/pull/731)?
- Is it hard coded or autogenerated as described in [the documentation](https://docs.flatpak.org/en/latest/python.html#building-multiple-python-dependencies)?
- What are the steps required to build a private puddletag flatpak? Will I find the required steps in the flatpak documentation, or are there some undocumented steps I should know about?

**Comment by raduzaharia-medium** [comment]
I didn't just do the research and development. It's all done. You can build the flatpak from my code [here](https://github.com/raduzaharia-medium/puddletag-flatpak). And it will fully work, editing from network shares included if you give the network permission via flatseal. [This](https://github.com/raduzaharia-medium/puddletag-flatpak/blob/main/dontusenativedialog.patch) is the patch. It simply applies a special option to all file dialogs in puddletag to allow compatibility with flatpak. 

Here are the steps to build the flatpak:
- download my github repository
- flatpak-builder build-dir net.puddletag.puddletag.yaml
- flatpak-builder --user --install --force-clean build-dir net.puddletag.puddletag.yaml
- flatpak run net.puddletag.puddletag

**Comment by eylenburg** [comment]
@raduzaharia-medium would you consider hosting your own Flatpak repo for Puddletag so that it's easier to install for "normal" users?

**Comment by lwbt** [comment]
@eylenburg In the mean time I suggest you take a look here: https://github.com/puddletag/puddletag/issues/773

I know it's not exactly what you are looking for, but very close to what the developers recommend and not for off from the Flatpak experience.

**Comment by raduzaharia-medium** [comment]
> @raduzaharia-medium would you consider hosting your own Flatpak repo for Puddletag so that it's easier to install for "normal" users?

After trying to host it on flathub (where this package belongs) and failing due to politics, I am way too disappointed to try anything else flatpak related. For a while I hosted the puddletag repo in my home network (it works excellent by the way) but then I just fell back to the usual RPM package for Fedora. Honestly? The fight is over stupid things and it's not worth it. Use the RPM/DEB package.

**Comment by lwbt** [comment]
> The fight is over stupid things and it's not worth it.

Again thanks for your efforts!

Hence my suggestion to look into pipx, not for the sake of more overhead and using another fancy tool, but for minimizing overhead from a consumer point of view, which we are. We don't enjoy activating a virtual environment before we can use the program.

Also waiting months on some distributions for some package to be picked up from upstream is not ideal and in light of immutable operating systems like Silverblue and Steam OS, options like Flatpak and programming language oriented repositories (PyPI, NPM, Go, Rust) or static builds seem to be the way to go forward. Hosting your own infrastructure is what should be done more often but comes with a a learning curve and a lot of frustration and shortage on time.

I'm grateful that Puddletag found some new maintainers, but they don't seem to be as constructive and pragmatic as other developers who publish their packages almost everywhere, because they realized that the landscape has changed a lot in the past 10 years.

**Comment by sandrotosi** [comment]
> I'm grateful that Puddletag found some new maintainers, but they don't seem to be as constructive and pragmatic as other developers who publish their packages almost everywhere, because they realized that the landscape has changed a lot in the past 10 years.

after this comment i'm going to lock this issue.

`pip install puddletag` just work.

im looking forward to your PRs to this project @lwbt , if you have something to contribute, constructively.
